File: 516.cpp
struct Solution {
    int longestPalindromeSubseq(const string &s) {
        vector<int> dp(s.size(), 0);
        int bottomleft, temp;
        dp[0] = 1;
        for (int j = 1; j < s.size(); ++j){
            bottomleft = dp[j];
            dp[j] = 1;
            for (int i = j - 1; i >= 0; --i){
                temp = bottomleft;
                bottomleft = dp[i];
                dp[i] = (s[i] == s[j] ? 2 + temp : max(dp[i], dp[i + 1]));
            }
        }
        return dp[0];
    }
};

File: sol/516.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        dp = [0] * len(s)
        dp[0] = 1
        for j in range(1, len(s)):
            bottomleft = dp[j]
            dp[j] = 1
            for i in range(j - 1, -1, -1):
                bottomleft, dp[i] = dp[i], (2 + bottomleft if s[i] == s[j] else max(dp[i], dp[i + 1]))
        return dp[0]


